File: t/CS/ImageProcessingDemo/UI/Command/DeskewExtendedDialog.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using Leadtools;
using Leadtools.ImageProcessing;
using Leadtools.ImageProcessing.Core;
using Leadtools.Demos; 

namespace ImageProcessingDemo
{
   //This dialog is used to allow the user to specify values which
   //will be used for the DeskewExtendedCommand

   public partial class DeskewExtendedDialog : Form
   {
      private DeskewExtendedCommand _DeskewExtendedCommand = null;
      private RasterColor _Color;
      private int _AngleRange;
      private RasterImage _image;

      public DeskewExtendedDialog(RasterImage image)
      {
         InitializeComponent();
         _DeskewExtendedCommand = new DeskewExtendedCommand();
         _Color = Leadtools.Demos.Converters.FromGdiPlusColor(Color.Black);
         _AngleRange = 2000;
         _image = image;
         _rbFast.Enabled = (_image.BitsPerPixel == 1);

         //Set command default values
         InitializeUI();
      }
      public DeskewExtendedDialog(DeskewExtendedCommand DeskewExtendedCommand)
      {
         InitializeComponent();
         _DeskewExtendedCommand = DeskewExtendedCommand;
         InitializeUI();
      }
      public DeskewExtendedCommand DeskewExtendedCommand
      {
         get
         {
            //Update command values
            UpdateCommand();
            return _DeskewExtendedCommand;
         }
         set
         {
            _DeskewExtendedCommand = value;
            InitializeUI();
         }
      }

      public RasterColor color
      {
         get
         {
            return _Color;
         }
         set
         {
            _Color = value;
         }
      }

      private void _btnOk_Click(object sender, EventArgs e)
      {
         UpdateCommand();
         this.DialogResult = DialogResult.OK;
         this.Close();
      }

      private void _btnCancel_Click(object sender, EventArgs e)
      {
         this.DialogResult = DialogResult.Cancel;
      }

      private void InitializeUI()
      {
         // Fill
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.FillExposedArea) == DeskewExtendedCommandFlags.FillExposedArea)
         {
            _cbFillExposedArea.Checked = true;
            _lblColor.BackColor = Leadtools.Demos.Converters.ToGdiPlusColor(_DeskewExtendedCommand.FillColor);
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(Color.Black);
            _lblColor.Enabled = true;
         }

         // Threshold
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.Threshold) == DeskewExtendedCommandFlags.Threshold)
         {
            _cbThreshold.Checked = true;
         }

         // Quality
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.RotateLinear) == DeskewExtendedCommandFlags.RotateLinear)
         {
            _rbLow.Checked = true;
         }
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.RotateResample) == DeskewExtendedCommandFlags.RotateResample)
         {
            _rbMedium.Checked = true;
         }
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.RotateBicubic) == DeskewExtendedCommandFlags.RotateBicubic)
         {
            _rbHigh.Checked = true;
         }

         // Type
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.DocumentImage) == DeskewExtendedCommandFlags.DocumentImage)
         {
            _rbTextOnly.Checked = true;
         }
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.DocumentAndTextImage) == DeskewExtendedCommandFlags.DocumentAndTextImage)
         {
            _rbTextImages.Checked = true;
         }

         // Speed
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.UseNormalRotate) == DeskewExtendedCommandFlags.UseNormalRotate)
         {
            _rbNormal.Checked = true;
         }
         if ((_DeskewExtendedCommand.Flags & DeskewExtendedCommandFlags.UseHighSpeedRotate) == DeskewExtendedCommandFlags.UseHighSpeedRotate)
         {
            _rbFast.Checked = true;
         }

         _txbAngleRange.Text = _AngleRange.ToString();
         _numAngleResolution.Value = 20;         
      }

      private void UpdateCommand()
      {
         _DeskewExtendedCommand.Flags = (DeskewExtendedCommandFlags)0;

         // Fill
         if (_cbFillExposedArea.Checked)
         {
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.FillExposedArea;
         }
         else
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.DoNotFillExposedArea;

         // Threshold
         if (_cbThreshold.Checked)
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.Threshold;
         else
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.NoThreshold;

         // Quality
         if (_rbLow.Checked)
         {
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.RotateLinear;
         }
         else if (_rbMedium.Checked)
         {
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.RotateResample;
         }
         else
         {
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.RotateBicubic;
         }

         // Type
         if (_rbTextOnly.Checked)
         {
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.DocumentImage;
         }
         else
         {
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.DocumentAndTextImage;
         }

         // Speed
         if (_rbNormal.Checked)
         {
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.UseNormalRotate;
         }
         else
         {
            _DeskewExtendedCommand.Flags |= DeskewExtendedCommandFlags.UseHighSpeedRotate;
         }

         _DeskewExtendedCommand.AngleRange = Convert.ToInt32(_txbAngleRange.Text);
         _DeskewExtendedCommand.AngleResolution = Convert.ToInt32(_numAngleResolution.Value);
      }

      private void _cbFillExposedArea_CheckedChanged(object sender, EventArgs e)
      {
         _lblColor.Enabled = _cbFillExposedArea.Checked;
      }

      private void lblColor_Click(object sender, EventArgs e)
      {
         ColorDialog dlg = new ColorDialog();
         dlg.Color = _lblColor.BackColor;
         if (dlg.ShowDialog() == DialogResult.OK)
         {
            _lblColor.BackColor = dlg.Color;
         }
      }

      private void _tbAngleRange_Scroll(object sender, EventArgs e)
      {
         _txbAngleRange.Text = _tbAngleRange.Value.ToString();         
      }

      private void _txbAngleRange_TextChanged(object sender, EventArgs e)
      {
         try
         {
            _txbAngleRange.Text = MainForm.IsValidNumber(_txbAngleRange.Text, 1, 4500);

            int Val = int.Parse(_txbAngleRange.Text);
            if (Val >= _tbAngleRange.Minimum && Val <= _tbAngleRange.Maximum)
               _tbAngleRange.Value = Val;

            if (_numAngleResolution.Value > Convert.ToInt32(_txbAngleRange.Text))
            {
               _numAngleResolution.Value = Convert.ToInt32(_txbAngleRange.Text);
            }
         }
         catch
         {
         }
      }

      private void _btnColor_Click(object sender, EventArgs e)
      {
         ColorDialog dlg = new ColorDialog();
         dlg.AllowFullOpen = true;
         dlg.AnyColor = true;
         DialogResult res = dlg.ShowDialog(this);
         if (res == DialogResult.OK)
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(dlg.Color);
         _lblColor.Refresh();
      }

      private void _lblColor_Paint(object sender, PaintEventArgs e)
      {
         e.Graphics.FillRectangle(new SolidBrush(Leadtools.Demos.Converters.ToGdiPlusColor(_Color)), _lblColor.ClientRectangle);
      }

      private void _numAngleResolution_Leave(object sender, EventArgs e)
      {
         DialogUtilities.NumericOnLeave(sender);
         if (_numAngleResolution.Value > Convert.ToInt32(_txbAngleRange.Text))
         {
            _numAngleResolution.Value = Convert.ToInt32(_txbAngleRange.Text);
         }
      }

      private void _numAngleResolution_ValueChanged(object sender, EventArgs e)
      {
         if (_numAngleResolution.Value > Convert.ToInt32(_txbAngleRange.Text))
         {
            _numAngleResolution.Value = Convert.ToInt32(_txbAngleRange.Text);
         }
      }
   }
}

File: t/VB/ImageProcessingDemo/UI/Command/DeskewExtendedDialog.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Imports Leadtools
Imports Leadtools.ImageProcessing
Imports Leadtools.ImageProcessing.Core

Namespace ImageProcessingDemo
   'This dialog is used to allow the user to specify values which
   'will be used for the DeskewExtendedCommand

   Partial Public Class DeskewExtendedDialog : Inherits Form
      Private _DeskewExtendedCommand As DeskewExtendedCommand = Nothing
      Private _Color As RasterColor
      Private _AngleRange As Integer
      Private _AngleResolution As Integer
      Private _Image As RasterImage
      
      Public Sub New(ByVal Image As RasterImage)
         InitializeComponent()
         _DeskewExtendedCommand = New DeskewExtendedCommand()
         _Color = Leadtools.Demos.Converters.FromGdiPlusColor(System.Drawing.Color.Black)
         _Image = Image

         If Not (_Image.BitsPerPixel = 1) Then
            _rbFast.Enabled = False
         End If

         _AngleRange = 2000
         _AngleResolution = 20
         InitializeUI()
      End Sub
      Public Sub New(ByVal DeskewExtendedCommand As DeskewExtendedCommand)
         InitializeComponent()
         _DeskewExtendedCommand = DeskewExtendedCommand
         InitializeUI()
      End Sub
      Public Property DeskewExtendedCommand() As DeskewExtendedCommand
         Get
            'Update command values
            UpdateCommand()
            Return _DeskewExtendedCommand
         End Get
         Set(ByVal value As DeskewExtendedCommand)
            _DeskewExtendedCommand = value
            InitializeUI()
         End Set
      End Property

      Public Property color() As RasterColor
         Get
            Return _Color
         End Get
         Set(ByVal value As RasterColor)
            _Color = value
         End Set
      End Property

      Private Sub _btnOk_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnOk.Click
         UpdateCommand()
         Me.DialogResult = DialogResult.OK
         Me.Close()
      End Sub

      Private Sub _btnCancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnCancel.Click
         Me.DialogResult = DialogResult.Cancel
      End Sub

      Private Sub InitializeUI()
         ' Fill

         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.FillExposedArea) = DeskewExtendedCommandFlags.FillExposedArea Then
            _cbFillExposedArea.Checked = True
            _lblColor.BackColor = Leadtools.Demos.Converters.ToGdiPlusColor(_DeskewExtendedCommand.FillColor)
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(System.Drawing.Color.Black)
            _lblColor.Enabled = True
         End If

         ' Threshold
         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.Threshold) = DeskewExtendedCommandFlags.Threshold Then
            _cbThreshold.Checked = True
         End If

         ' Quality
         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.RotateLinear) = DeskewExtendedCommandFlags.RotateLinear Then
            _rbLow.Checked = True
         End If
         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.RotateResample) = DeskewExtendedCommandFlags.RotateResample Then
            _rbMedium.Checked = True
         End If
         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.RotateBicubic) = DeskewExtendedCommandFlags.RotateBicubic Then
            _rbHigh.Checked = True
         End If

         ' Type
         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.DocumentImage) = DeskewExtendedCommandFlags.DocumentImage Then
            _rbTextOnly.Checked = True
         End If
         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.DocumentAndTextImage) = DeskewExtendedCommandFlags.DocumentAndTextImage Then
            _rbTextImages.Checked = True
         End If

         ' Speed
         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.UseNormalRotate) = DeskewExtendedCommandFlags.UseNormalRotate Then
            _rbNormal.Checked = True
         End If
         If (_DeskewExtendedCommand.Flags And DeskewExtendedCommandFlags.UseHighSpeedRotate) = DeskewExtendedCommandFlags.UseHighSpeedRotate Then
            _rbFast.Checked = True
         End If

         _txbAngleRange.Text = _AngleRange.ToString()
         _txtAngleResolution.Text = _AngleResolution.ToString()
      End Sub

      Private Sub UpdateCommand()
         _DeskewExtendedCommand.Flags = CType(0, DeskewExtendedCommandFlags)

         ' Fill
         If _cbFillExposedArea.Checked Then
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.FillExposedArea
         Else
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.DoNotFillExposedArea
         End If

         ' Threshold
         If _cbThreshold.Checked Then
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.Threshold
         Else
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.NoThreshold
         End If

         ' Quality
         If _rbLow.Checked Then
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.RotateLinear
         ElseIf _rbMedium.Checked Then
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.RotateResample
         Else
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.RotateBicubic
         End If

         ' Type
         If _rbTextOnly.Checked Then
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.DocumentImage
         Else
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.DocumentAndTextImage
         End If

         ' Speed
         If _rbNormal.Checked Then
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.UseNormalRotate
         Else
            _DeskewExtendedCommand.Flags = _DeskewExtendedCommand.Flags Or DeskewExtendedCommandFlags.UseHighSpeedRotate
         End If

         _DeskewExtendedCommand.AngleRange = Convert.ToInt32(_txbAngleRange.Text)
         _DeskewExtendedCommand.AngleResolution = Convert.ToInt32(_txtAngleResolution.Text)
      End Sub

      Private Sub _cbFillExposedArea_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _cbFillExposedArea.CheckedChanged
         _lblColor.Enabled = _cbFillExposedArea.Checked
      End Sub

      Private Sub lblColor_Click(ByVal sender As Object, ByVal e As EventArgs)
         Dim dlg As ColorDialog = New ColorDialog()
         dlg.Color = _lblColor.BackColor
         If dlg.ShowDialog() = DialogResult.OK Then
            _lblColor.BackColor = dlg.Color
         End If
      End Sub

      Private Sub _tbAngleRange_Scroll(ByVal sender As Object, ByVal e As EventArgs) Handles _tbAngleRange.Scroll
         _txbAngleRange.Text = _tbAngleRange.Value.ToString()
      End Sub

      Private Sub _txbAngleRange_TextChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _txbAngleRange.TextChanged
         Try
            _txbAngleRange.Text = MainForm.IsValidNumber(_txbAngleRange.Text, 1, 4500)

            Dim Val As Integer = Integer.Parse(_txbAngleRange.Text)
            If Val >= _tbAngleRange.Minimum AndAlso Val <= _tbAngleRange.Maximum Then
               _tbAngleRange.Value = Val
            End If

            If Convert.ToInt32(_txtAngleResolution.Text) > Convert.ToInt32(_txbAngleRange.Text) Then
               _txtAngleResolution.Text = _txbAngleRange.Text
            End If
         Catch
         End Try
      End Sub

      Private Sub _btnColor_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnColor.Click
         Dim dlg As ColorDialog = New ColorDialog()
         dlg.AllowFullOpen = True
         dlg.AnyColor = True
         Dim res As DialogResult = dlg.ShowDialog(Me)
         If res = DialogResult.OK Then
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(dlg.Color)
         End If
         _lblColor.Refresh()
      End Sub

      Private Sub _lblColor_Paint(ByVal sender As Object, ByVal e As PaintEventArgs) Handles _lblColor.Paint
         e.Graphics.FillRectangle(New SolidBrush(Leadtools.Demos.Converters.ToGdiPlusColor(_Color)), _lblColor.ClientRectangle)
      End Sub

      Private Sub _txtAngleResolution_TextChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles _txtAngleResolution.TextChanged
         Try
            _txtAngleResolution.Text = MainForm.IsValidNumber(_txtAngleResolution.Text, 1, 4500)

            Dim Val As Integer = Integer.Parse(_txbAngleRange.Text)
            If Val >= _tbAngleRange.Minimum AndAlso Val <= _tbAngleRange.Maximum Then
               _tbAngleRange.Value = Val
            End If

            If Convert.ToInt32(_txtAngleResolution.Text) > Convert.ToInt32(_txbAngleRange.Text) Then
               _txtAngleResolution.Text = _txbAngleRange.Text
            End If
         Catch
         End Try
      End Sub

      Private Sub _txtAngleResolution_Leave(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles _txtAngleResolution.Leave
         If Convert.ToInt32(_txtAngleResolution.Text) > Convert.ToInt32(_txbAngleRange.Text) Then
            _txtAngleResolution.Text = _txbAngleRange.Text
         End If
      End Sub
   End Class
End Namespace
